[ iOS ] http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html is a flakey failure (297081)

pooneh-nb opened the pull request:
#### b2427a37d4827f4174ac35193f7c94fa1b6fec05
<pre>
[ iOS ] http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html is a flakey failure (297081)
<a href="https://bugs.webkit.org/show_bug.cgi?id=304207">https://bugs.webkit.org/show_bug.cgi?id=304207</a>
<a href="https://rdar.apple.com/157787103">rdar://157787103</a>

Reviewed by Matthew Finkel.

Radar <a href="https://rdar.apple.com/157787103">rdar://157787103</a> reported http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html as flaky on all iOS configurations.
Step 1 was flaky due to a race condition, so I added a setTimeout-based timer to avoid the race and stabilize the test.

* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html:
* LayoutTests/platform/ios/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/304871@main">https://commits.webkit.org/304871@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e47c86673830ccb43d5ead89c2609feeab858fd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136332 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8689 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47612 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144044 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89303 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dd9af3b6-27a2-4cf5-91da-a65aff8c7593) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9370 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8533 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104243 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/74869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/00f2fad4-0a35-41d3-b6be-b51461360943) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139277 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6820 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122164 "Found 1 new API test failure: TestWebKitAPI.MessagePort.MessageToClosedPort (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85076 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b15b4250-e573-4a47-83ca-61e2941d0eb3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6473 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4138 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4635 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115766 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40364 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146788 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8371 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40932 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112584 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8388 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7034 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112928 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28765 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6404 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118468 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62358 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8419 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36525 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8137 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8359 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8211 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -8203,7 +8203,7 @@ webkit.org/b/297056 http/tests/site-isolation/touch-events/touch-event-coordinat
 webkit.org/b/297067 fast/forms/state-restore-form-associated-custom-elements.html [ Pass Timeout ]
 
-webkit.org/b/297081 http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html [ Pass Failure ]
+# webkit.org/b/297081 http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html [ Pass Failure ]
```
Shouldn't this be removed instead?

johnwilander commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -614,7 +614,26 @@ void NetworkStorageSession::deleteCookies(const ClientOrigin& origin, Completion
+        (int)enabled);
+
+    return enabled;
+    // return thirdPartyCookieBlockingDecisionForRequest(firstParty, url, frameID, pageID, shouldRelaxThirdPartyCookieBlocking, isKnownCrossSiteTracker) != ThirdPartyCookieBlockingDecision::All;
```
If you intended to make the above change, this comment should be removed.

johnwilander commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -614,7 +614,26 @@ void NetworkStorageSession::deleteCookies(const ClientOrigin& origin, Completion
+
+    bool enabled = decision != ThirdPartyCookieBlockingDecision::All;
+
+    WTFLogAlways("NetworkStorageSession::cookiesEnabled pid=%d mode=%d t=%.6f fp=%s url=%s page=%s frame=%s relax=%d known=%d decision=%d enabled=%d",
```
Is this just from your debugging? Maybe you didn't intend to change this function at all?

johnwilander commented on LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html:
```diff
@@ -36,6 +39,9 @@
             testRunner.setManagedDomains([ "http://127.0.0.1:8000" ], () => {
+                // setTimeout(() => {
+                //     iframeElement.contentWindow.postMessage("getCookieEnabled", "*");
+                // }, 0);
```
This looks like debugging code.

johnwilander commented on LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html:
```diff
@@ -28,6 +28,9 @@
     function runTest() {
         switch(currentStep) {
         case 1:
+            // sanity check: 
```
It's unclear how this checks sanity. Could we be more clear as to what is checked (if we keep the comment, that is)?

johnwilander commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2428,11 +2428,22 @@ void WebProcess::displayDidRefresh(uint32_t displayID, const DisplayUpdate& disp
 void WebProcess::setThirdPartyCookieBlockingMode(ThirdPartyCookieBlockingMode thirdPartyCookieBlockingMode, CompletionHandler<void()>&& completionHandler)
 {
+    auto t = MonotonicTime::now().secondsSinceEpoch().value();
+    WTFLogAlways("WebProcess::setThirdPartyCookieBlockingMode ENTER pid=%d t=%.6f current3PBlockingMode=%d incoming3PBlockingMode=%d",
```
Is this new logging intentional?

johnwilander commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2428,11 +2428,22 @@ void WebProcess::displayDidRefresh(uint32_t displayID, const DisplayUpdate& disp
+        getpid(), t, (int)m_thirdPartyCookieBlockingMode, (int)thirdPartyCookieBlockingMode);
+
     if (m_thirdPartyCookieBlockingMode != thirdPartyCookieBlockingMode) {
+        WTFLogAlways("WebProcess::setThirdPartyCookieBlockingMode APPLIED pid=%d t=%.6f old=%d new=%d",
```
Ditto

johnwilander commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2428,11 +2428,22 @@ void WebProcess::displayDidRefresh(uint32_t displayID, const DisplayUpdate& disp
         m_thirdPartyCookieBlockingMode = thirdPartyCookieBlockingMode;
-        if (m_thirdPartyCookieBlockingMode != ThirdPartyCookieBlockingMode::All)
+        if (m_thirdPartyCookieBlockingMode != ThirdPartyCookieBlockingMode::All) {
+            WTFLogAlways("WebProcess::setThirdPartyCookieBlockingMode updateCachedCookiesEnabled pid=%d", getpid());
```
Ditto

johnwilander commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2428,11 +2428,22 @@ void WebProcess::displayDidRefresh(uint32_t displayID, const DisplayUpdate& disp
             updateCachedCookiesEnabled();
+        }
     }
+     WTFLogAlways("WebProcess::setThirdPartyCookieBlockingMode ACK pid=%d t=%.6f mode=%d",
```
Ditto

johnwilander reviewed (commented):
I don't see a functional change or a change in test expectations. Maybe I'm missing it.

Commit:
[ iOS ] http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html is a flakey failure (297081)
https://bugs.webkit.org/show_bug.cgi?id=304207
rdar://157787103

Reviewed by Matthew Finkel.

Radar rdar://157787103 reported http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html as flaky on all iOS configurations.
Step 1 was flaky due to a race condition, so I added a setTimeout-based timer to avoid the race and stabilize the test.

* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html:
* LayoutTests/platform/ios/TestExpectations:

Canonical link: https://commits.webkit.org/304871@main
